File: samples/nodejs-typescript-food-catalog/src/functions/queue-connection.ts
import { app, InvocationContext } from "@azure/functions";
import { ResponseType } from "@microsoft/microsoft-graph-client";
import { ExternalConnectors } from "@microsoft/microsoft-graph-types";
import { config } from "../common/config";
import { client } from "../common/graphClient";
import { enqueueCheckStatus, startCrawl } from "../common/queueClient";
import { resultLayout } from "../common/resultLayout";
import { ConnectionMessage } from "../common/ConnectionMessage";

async function createConnection(connectorId: string, connectorTicket: string) {
    const { id, name, description, activitySettings, searchSettings } = config.connector;
    searchSettings.searchResultTemplates[0].layout = resultLayout;

    await client
        .api('/external/connections')
        .version('beta')
        .header('GraphConnectors-Ticket', connectorTicket)
        .post({
            id,
            connectorId,
            name,
            description,
            activitySettings,
            searchSettings
        });
}

async function createSchema() {
    const { id, schema } = config.connector;
    const res: Response = await client
        .api(`/external/connections/${id}/schema`)
        .responseType(ResponseType.RAW)
        .header('content-type', 'application/json')
        .patch({
            baseType: 'microsoft.graph.externalItem',
            properties: schema
        });

    const location: string = res.headers.get('Location');
    await enqueueCheckStatus(location);
}

async function checkSchemaStatus(location: string, context: InvocationContext) {
    const res: ExternalConnectors.ConnectionOperation = await client
        .api(location)
        .get();

    context.log(`Schema provisioning status: ${res.status}`);

    switch (res.status) {
        case 'inprogress':
            await enqueueCheckStatus(location);
            break;
        case 'completed':
            await startCrawl('full');
            break;
    }
}

async function deleteConnection() {
    await client.api(`/external/connections/${config.connector.id}`).delete();
}

app.storageQueue("connectionQueue", {
    connection: "AzureWebJobsStorage",
    queueName: "queue-connection",
    handler: async (message: ConnectionMessage, context: InvocationContext) => {
        context.log('Received message from queue queue-connection');
        context.log(JSON.stringify(message, null, 2));

        const { action, connectorId, connectorTicket, location } = message;

        switch (action) {
            case 'create':
                context.log('Creating connection...');
                await createConnection(connectorId, connectorTicket);
                context.log('Connection created');
                context.log('Submitting schema for provisioning...');
                createSchema();
                context.log('Schema submitted');
                break;
            case 'delete':
                context.log('Deleting connection...');
                await deleteConnection();
                context.log('Connection deleted');
                break;
            case 'status':
                context.log('Checking schema status...');
                await checkSchemaStatus(location, context);
                context.log('Schema status checked');
                break;
            default:
                break;
        }
    }
})

File: samples/dotnet-csharp-graphdocs-ttk/Functions/QueueConnection.cs
using System.Text.Json;
using Azure.Storage.Queues;
using Azure.Storage.Queues.Models;
using GraphDocsConnector.Messages;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;
using Microsoft.Graph;
using Microsoft.Graph.Models.ExternalConnectors;

namespace GraphDocsConnector.Functions
{
    public class QueueConnection
    {
        private readonly ILogger<QueueConnection> _logger;
        private readonly GraphServiceClient _graphClient;
        private readonly QueueClient _queueConnectionClient;
        private readonly QueueClient _queueContentClient;

        public QueueConnection(QueueServiceClient queueClient, GraphServiceClient graphClient, ILogger<QueueConnection> logger)
        {
            _queueConnectionClient = queueClient.GetQueueClient("queue-connection");
            _queueContentClient = queueClient.GetQueueClient("queue-content");
            _graphClient = graphClient;
            _logger = logger;
        }

        [Function(nameof(QueueConnection))]
        public async Task Run([QueueTrigger("queue-connection", Connection = "AzureWebJobsStorage")] QueueMessage message)
        {
            var connectionMessage = JsonSerializer.Deserialize<ConnectionMessage>(message.MessageText);
            if (connectionMessage is null)
            {
                _logger.LogError($"Couldn't deserialize message: {message.MessageText}");
                return;
            }

            switch (connectionMessage.Action)
            {
                case ConnectionMessageAction.Create:
                    _logger.LogInformation("Creating connection...");
                    await CreateConnection(connectionMessage);
                    _logger.LogInformation("Connection created");
                    _logger.LogInformation("Submitting schema for provisioning...");
                    await CreateSchema();
                    _logger.LogInformation("Schema submitted");
                    break;
                case ConnectionMessageAction.Delete:
                    _logger.LogInformation("Deleting connection...");
                    await DeleteConnection();
                    _logger.LogInformation("Connection deleted");
                    break;
                case ConnectionMessageAction.Status:
                    _logger.LogInformation("Checking schema status...");
                    await CheckSchemaStatus(connectionMessage.Location!);
                    _logger.LogInformation("Schema status checked");
                    break;
            }
        }

        private async Task CheckSchemaStatus(string location)
        {
            var res = await _graphClient.External
                .Connections[ConnectionConfiguration.ExternalConnection.Id]
                // intentionally empty because we're getting the operation ID
                // from the location using the WithUrl construct
                .Operations[""]
                .WithUrl(location).GetAsync();

            if (res is null)
            {
                _logger.LogError($"Operation status for {location} is null");
                return;
            }

            _logger.LogInformation($"Schema provisioning status: {res.Status}");

            switch (res.Status)
            {
                case ConnectionOperationStatus.Inprogress:
                    Queue.EnqueueCheckStatus(_queueConnectionClient, location);
                    break;
                case ConnectionOperationStatus.Completed:
                    Queue.StartCrawl(_queueContentClient, CrawlType.Full);
                    break;
                default:
                    _logger.LogWarning($"Unsupported status");
                    break;
            }
        }

        private async Task CreateConnection(ConnectionMessage connectionMessage)
        {
            var externalConnection = ConnectionConfiguration.ExternalConnection;
            externalConnection.ConnectorId = connectionMessage.ConnectorId;
            
            await _graphClient.External.Connections
                .PostAsync(externalConnection, request =>
                {
                    request.Headers.Add("GraphConnectors-Ticket", connectionMessage.ConnectorTicket!);
                });
        }

        private async Task CreateSchema()
        {
            var schemaRequest = _graphClient.External
                .Connections[ConnectionConfiguration.ExternalConnection.Id]
                .Schema
                .ToPatchRequestInformation(ConnectionConfiguration.Schema);
            var httpRequestMessage = await _graphClient.RequestAdapter
                .ConvertToNativeRequestAsync<HttpRequestMessage>(schemaRequest);
            if (httpRequestMessage is null)
            {
                _logger.LogError("httpRequestMessage is null");
                return;
            }

            var httpClient = Utils.GetHttpClient();
            var res = await httpClient.SendAsync(httpRequestMessage);
            var location = res.Headers.GetValues("location")?.FirstOrDefault();
            if (string.IsNullOrEmpty(location))
            {
                _logger.LogError("Schema operation status location is empty");
                return;
            }

            Queue.EnqueueCheckStatus(_queueConnectionClient, location);
        }

        private async Task DeleteConnection()
        {
            await _graphClient.External
                .Connections[ConnectionConfiguration.ExternalConnection.Id]
                .DeleteAsync();
        }
    }
}
